### ISSUE TITLE
Particles + mesh deformation hits artificial cell assert

### ISSUE THREAD

**Issue by anne-glerum** [geodynamics/aspect#4023]
When running a model with particles and mesh deformation in parallel and in debug mode (assert is not triggered in release), the following error is triggered:
```
An error occurred in line <3271> of file </scratch/usr/bbpanneg/software/dealii/candi_9.2.0/tmp/unpack/deal.II-v9.2.0/include/deal.II/dofs/dof_accessor.templates.h> in function
    void dealii::DoFCellAccessor<DoFHandlerType, lda>::get_dof_values(const InputVector&, ForwardIterator, ForwardIterator) const [with InputVector = dealii::TrilinosWrappers::MPI::Vector; ForwardIterator = double*; DoFHandlerType = dealii::DoFHandler<2, 2>; bool level_dof_access = false]

 The violated condition was: 
    this->is_artificial() == false
 Additional information: 
    Can't ask for DoF indices on artificial cells.

Stacktrace:
-----------
#0  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: void dealii::DoFCellAccessor<dealii::DoFHandler<2, 2>, false>::get_dof_values<dealii::TrilinosWrappers::MPI::Vector, double*>(dealii::TrilinosWrappers::MPI::Vector const&, double*, double*) const
#1  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: void dealii::DoFCellAccessor<dealii::DoFHandler<2, 2>, false>::get_dof_values<dealii::TrilinosWrappers::MPI::Vector, double>(dealii::TrilinosWrappers::MPI::Vector const&, dealii::Vector<double>&) const
#2  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: dealii::MappingQ1Eulerian<2, dealii::TrilinosWrappers::MPI::Vector, 2>::get_vertices(dealii::TriaIterator<dealii::CellAccessor<2, 2> > const&) const
#3  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: dealii::MappingQ1Eulerian<2, dealii::TrilinosWrappers::MPI::Vector, 2>::compute_mapping_support_points(dealii::TriaIterator<dealii::CellAccessor<2, 2> > const&) const
#4  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: dealii::MappingQGeneric<2, 2>::transform_real_to_unit_cell(dealii::TriaIterator<dealii::CellAccessor<2, 2> > const&, dealii::Point<2, double> const&) const
#5  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0:
#6  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: std::pair<dealii::Triangulation<2, 2>::active_cell_iterator, dealii::Point<2, double> > dealii::GridTools::find_active_cell_around_point<2, dealii::Triangulation, 2>(dealii::Mapping<2, 2> const&, dealii::Triangulation<2, 2> const&, dealii::Point<2, double> const&, std::vector<bool, std::allocator<bool> > const&)
#7  /scratch/usr/bbpanneg/software/dealii/candi_9.2.0/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: dealii::Particles::ParticleHandler<2, 2>::sort_particles_into_subdomains_and_cells()
#8  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: aspect::Particle::World<2>::advect_particles()  
#9  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: aspect::Particle::World<2>::advance_timestep()
#10  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: aspect::Simulator<2>::assemble_and_solve_composition(bool, std::vector<double, std::allocator<double> >*)
#11  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: aspect::Simulator<2>::solve_iterated_advection_and_stokes()
#12  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: aspect::Simulator<2>::solve_timestep()
#13  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: aspect::Simulator<2>::run()
#14  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: void run_simulator<2>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
#15  /scratch/usr/bbpanneg/software/aspect/build_debug__melt_visco_plastic_PM01_2/aspect: main
```

This seems very similar to #2050.

**Comment by gassmoeller** [comment]
Hi Anne, Yes it is a similar issue. In essence the particle handler tries to find the new cell for a particle among the neighbor cells. When it does not succeed it tries to search all cells, and if that fails as well it deletes the particle. It expects (and catches) a certain type of exception if the particle is outside of the domain to determine when to delete the particle. The problem seems to arise if the particle is in an artificial cell, because then the search function throws a different exception than the one the algorithm expects. Can you try modifying line 1021 (assuming deal.II 9.2.0) of source/particles/particle_handler from:

```
              catch (GridTools::ExcPointNotFound<spacedim> &)
``` 
to
```
              catch (...)
```

and check if that fixes the issue? We will need a better fix than that, but it would tell me if that is indeed the issue.

**Comment by anne-glerum** [comment]
Hi Rene, this doesn't fix the issue, this particular catch is not caught actually. I could look into it more later.

**Comment by tjhei** [comment]
This might be an Assert and not an AssertThrow, which means there is no way to catch it. 😢
This might need a Bugfix in deal.Ii...

**Comment by gassmoeller** [comment]
Hi Anne, I have tried a number of approaches to reproduce your issue, but could not trigger it. Do you have a simple test case?
Also could you try if this still happens with the current deal.II development version? There have been a number of changes in the deal.II functions, and I find it hard to see which issue could cause this, and if it is still present.

**Comment by anne-glerum** [comment]
Hi Rene, thank you for looking into it! I'll try with the deal.II dev version first, and then try to abstract a simpler test case if necessary.

**Comment by anne-glerum** [comment]
Hi Rene, an ASPECT dev install against deal.II dev fails at the moment because ParticleHandler function `n_locally_owned_particles()` has changed name (`n_locally_owned_particles_ids()`). I'll make a PR to fix it.

**Comment by gassmoeller** [comment]
Thanks, I think that was caused by dealii/dealii#12119

**Comment by gassmoeller** [comment]
@anne-glerum: Better wait with the PR, deal.II will revert that change (it was included in another change accidentally). You could also take a deal.II version that is just a week old or so (before PR 12119 was merged).

**Comment by peterrum** [comment]
@anne-glerum @gassmoeller Sorry for the unintended change. I hope the patch should be merged in the next hours.

**Comment by anne-glerum** [comment]
@peterrum Thanks for the quick fix!

@gassmoeller With the patched deal.II, it seems runs using more than 1 node crash with this error:
```
380 terminate called after throwing an instance of 'std::system_error'
381   what():  Resource temporarily unavailable
```
I'm running a small model now to see how the second restart goes. Also, I will check with deal.II 9.2 whether I can still use multiple nodes or whether there is something else wrong.

**Comment by gassmoeller** [comment]
Hmm, that is a weird error. Are you sure you recompiled aspect and any plugins that you use against the recent deal.II?

I looked through the history of the relevant function in deal.II, and I am reasonably sure that https://github.com/dealii/dealii/pull/10540/ fixed this error, so it should already be included in deal.II 9.3.0.

**Comment by anne-glerum** [comment]
Yes I recompiled aspect and the shared libs. For 2 procs this is the header:
```
      3 --     . version 2.3.0-pre (melt_visco_plastic_PM01, 85217e3)
      4 --     . using deal.II 9.3.0-pre (master, a621b86)
      5 --     .       with 32 bit indices and vectorization level 3 (512 bits)
      6 --     . using Trilinos 12.18.1
      7 --     . using p4est 2.2.0
      8 --     . running in DEBUG mode
      9 --     . running with 2 MPI processes
```
so that's deal.II from this morning. For 384 procs, I just get the system_error, not even the header is outputted. The same model with deal.II 9.2.0 and 384 procs runs.

**Comment by anne-glerum** [comment]
I've made a completely fresh install of deal.II master and aspect (master from May 3). Running a cookbook model with 48 procs crashes with std::system_error, with 4 procs doesn't. main{} is not reached before the crash; the parallel debugger also doesn't tell me anything (same error). I guess there is something wrong with my setup, or has anybody encountered the same error?

**Comment by anne-glerum** [comment]
With the fix in #4040, and deal.II 9.3 I no longer run into this issue.
